packages/flutter/lib/src/painting/text_painter.dart

```diff
@@ -270,17 +270,179 @@ class _UntilTextBoundary extends TextBoundary {
+        return clampDouble(layout.longestLine, minWidth, maxWidth);
+      case TextWidthBasis.parent:
+        return clampDouble(layout.maxIntrinsicLineExtent, minWidth, maxWidth);
+    }
+  }
+  // Try to resize the contentWidth to fit the new input constraints, by just
+  // adjusting the paint offset (so no line-breaking changes needed).
+  //
+  // Returns false if the new constraints require re-computing the line breaks,
+  // in which case no side effects will occur.
+  bool _resizeToFit(double minWidth, double maxWidth, TextWidthBasis widthBasis) {
+    assert(layout.maxIntrinsicLineExtent.isFinite);
+    // The assumption here is that if a Paragraph's width is already >= its
+    // maxIntrinsicWidth, further increasing the input width does not change its
+    // layout (but may change the paint offset if it's not left-aligned). This is
+    // true even for TextAlign.justify: when width >= maxIntrinsicWidth
+    // TextAlign.justify will behave exactly the same as TextAlign.start.
+    //
+    // An exception to this is when the text is not left-aligned, and the input
+    // width is double.infinity. Since the resulting Paragraph will have a width
+    // of double.infinity, and to make the text visible the paintOffset.dx is
+    // bound to be double.negativeInfinity, which invalidates all arithmetic
+    // operations.
+    final double newContentWidth = _contentWidthFor(minWidth, maxWidth, widthBasis, layout);
+    if (newContentWidth == contentWidth) {
+      return true;
+    }
+    assert(minWidth <= maxWidth);
+    // Always needsLayout when the current paintOffset and the paragraph width are not finite.
+    if (!paintOffset.dx.isFinite && !paragraph.width.isFinite && minWidth.isFinite) {
+      assert(paintOffset.dx == double.infinity);
+      assert(paragraph.width == double.infinity);
+      return false;
+    }
+    final double maxIntrinsicWidth = layout._paragraph.maxIntrinsicWidth;
+    if ((layout._paragraph.width - maxIntrinsicWidth) > -precisionErrorTolerance && (maxWidth - maxIntrinsicWidth) > -precisionErrorTolerance) {
+      // Adjust the paintOffset and contentWidth to the new input constraints.
+      contentWidth = newContentWidth;
+      return true;
+    }
+    return false;
+  }
+
+  // ---- Cached Values ----
+
+  List<TextBox>? _inlinePlaceholderBoxes;
+  List<ui.LineMetrics>? _lineMetricsCache;
+  // Holds the TextPosition the last caret metrics were computed with. When new
+  // values are passed in, we recompute the caret metrics only as necessary.
+  TextPosition? _previousCaretPosition;
+}
+
 /// This is used to cache and pass the computed metrics regarding the
 /// caret's size and position. This is preferred due to the expensive
 /// nature of the calculation.
-class _CaretMetrics {
-  const _CaretMetrics({required this.offset, this.fullHeight});
+// This should be a sealed class: A _CaretMetrics is either a _LineCaretMetrics
+// or an _EmptyLineCaretMetrics.
+class _CaretMetrics { }
```

Comment:
Will `_CaretMetrics` ever be instantiated? If not, maybe make it an abstract class and have child classes extend it?